### PR TITLE
Align assignAllIds with documentation

### LINKED ISSUE #1150: Annotator assignAllIds misleading
The documentation for the Annotator assignAllIds method is erroneous.
The return value does not indicate if the model has been modified.
The code itself for assignAllIds is, how to say, less than the desired standard.

The first question is, do we want the method to do what the method documentation states?
The second question, do we want to refactor the method as part of this issue?

### PR DESCRIPTION
**Pull request by hsorby**
Yes, we do want the code to do what the function documentation says.
Yes, we do want to improve the codebase when we notice less than optimal sections where changes are being made.

Fixes #1150.

### CODE REVIEW

File: src/annotator.cpp [hunk 1]
```diff
@@ -812,30 +827,30 @@ bool Annotator::assignAllIds(ModelPtr &model)
         issue->mPimpl->setReferenceRule(Issue::ReferenceRule::ANNOTATOR_NULL_MODEL);
         return false;
     }
+    auto curModel = this->model();
```
Comment (nickerso) [inline src/annotator.cpp]: Documentation says that this should overwrite any existing model in this Annotator. I think we should stick with the documented approach and not cache the current model.

File: tests/annotator/annotator.cpp [hunk 2]
```diff
@@ -1598,11 +1622,23 @@ TEST(Annotator, assignAllIdsNonblankModel)
...
 
+TEST(Annotator, assignAllIdsDuplicateIdsModel)
+{
+    auto parser = libcellml::Parser::create();
+    auto printer = libcellml::Printer::create();
+    auto model = parser->parseModel(modelStringLotsOfDuplicateIds);
+    auto annotator = libcellml::Annotator::create();
+
+    EXPECT_FALSE(annotator->assignAllIds(model));
+    EXPECT_EQ(size_t(0), annotator->errorCount());
```
Comment (nickerso) [inline tests/annotator/annotator.cpp:1641]: if `assignAllIds(model)` goes back to not caching the current model, then this should be non-zero...

**Review by nickerso — CHANGES_REQUESTED** [review]

File: tests/annotator/annotator.cpp (hunk 2 above)
Comment (hsorby) [reply tests/annotator/annotator.cpp:1641]: This won't happen, the annotator doesn't raise errors if there are duplicate ids.  This question can be answered by the Annotator but the Validator is the tool for reporting this problem.  As far as the annotator is concerned if the user wants to set a duplicate id on an item who are we to stop them.

File: src/annotator.cpp (hunk 1 above)
Comment (hsorby) [reply src/annotator.cpp]: [d32b338](https://github.com/cellml/libcellml/pull/1151/commits/d32b3387c5ceba5aacc8a923de17f3db67f6062a)

**Review by nickerso — DISMISSED** [review]

**Review by nickerso — DISMISSED** [review]

File: src/annotator.cpp [hunk 3]
```diff
@@ -842,245 +853,159 @@ bool Annotator::assignIds(CellmlElementType type)
...
         default: /* case CellmlElementType::UNDEFINED */
             changed = false;
             break;
         }
         setModel(model);
+        changed = pFunc()->idCount() > initialSize;
     } else {
         pFunc()->addIssueNoModel();
     }
     return changed;
```
Comment (agarny) [inline src/annotator.cpp]: I would replace the this snippet of code with:
```c++
    return pFunc()->idCount() > initialSize;
}

pFunc()->addIssueNoModel();

return true;
```
and remove previous references to `changed`.

File: src/annotator.cpp [hunk 4]
```diff
@@ -842,245 +853,159 @@ bool Annotator::assignIds(CellmlElementType type)
...
-        if (importSource->id().empty()) {
-            auto id = makeUniqueId();
-            importSource->setId(id);
-            auto entry = AnyCellmlElement::AnyCellmlElementImpl::create();
-            entry->mPimpl->setImportSource(importSource);
-            mIdList.insert(std::make_pair(id, convertToWeak(entry)));
-        }
-    }
+    bool inHierarchy = std::dynamic_pointer_cast<libcellml::Model>(component->parent()) == nullptr || component->componentCount() > 0;
+    return (type == CellmlElementType::COMPONENT_REF || all) && component->encapsulationId().empty() && inHierarchy;
```
Comment (agarny) [inline src/annotator.cpp]: Parentheses and formatting:
```c++
bool inHierarchy = std::dynamic_pointer_cast<libcellml::Model>((component->parent()) == nullptr)
                   || (component->componentCount() > 0);

return ((type == CellmlElementType::COMPONENT_REF) || all)
       && component->encapsulationId().empty()
       && inHierarchy;
```

File: src/annotator.cpp [hunk 5]
```diff
@@ -842,245 +853,159 @@ bool Annotator::assignIds(CellmlElementType type)
...
-    for (size_t u = 0; u < model->unitsCount(); ++u) {
-        auto us = model->units(u);
-        if (us->id().empty()) {
-            auto id = makeUniqueId();
-            us->setId(id);
-            auto entry = AnyCellmlElement::AnyCellmlElementImpl::create();
-            entry->mPimpl->setUnits(us);
-            mIdList.insert(std::make_pair(id, convertToWeak(entry)));
-        }
+    if ((type == CellmlElementType::COMPONENT || all) && component->id().empty()) {
```
Comment (agarny) [inline src/annotator.cpp]:
```c++
if (((type == CellmlElementType::COMPONENT) || all)
    && component->id().empty()) {
```

File: src/annotator.cpp [hunk 6]
```diff
@@ -842,245 +853,159 @@ bool Annotator::assignIds(CellmlElementType type)
...
-        for (size_t i = 0; i < us->unitCount(); i++) {
-            if (us->unitId(i).empty()) {
+    if (assignEncapsulationId(component, type, all)) {
+        auto id = makeUniqueId();
+        component->setEncapsulationId(id);
+        auto entry = AnyCellmlElement::AnyCellmlElementImpl::create();
+        entry->mPimpl->setComponentRef(component);
+        mIdList.insert(std::make_pair(id, convertToWeak(entry)));
+    }
+    if (type == CellmlElementType::VARIABLE || all) {
```
Comment (agarny) [inline src/annotator.cpp]:
```c++
if ((type == CellmlElementType::VARIABLE) || all) {
```

File: src/annotator.cpp [hunk 7]
```diff
@@ -842,245 +853,159 @@ bool Annotator::assignIds(CellmlElementType type)
...
-        auto id = makeUniqueId();
-        parent->setId(id);
-        auto entry = AnyCellmlElement::AnyCellmlElementImpl::create();
-        entry->mPimpl->setComponent(parent);
-        mIdList.insert(std::make_pair(id, convertToWeak(entry)));
+    for (size_t vIndex = 0; vIndex < component->variableCount(); ++vIndex) {
+        auto v1 = component->variable(vIndex);
+        for (size_t eIndex = 0; eIndex < v1->equivalentVariableCount(); ++eIndex) {
+            auto v2 = v1->equivalentVariable(eIndex);
+            if ((type == CellmlElementType::CONNECTION || all) && Variable::equivalenceConnectionId(v1, v2).empty()) {
```
Comment (agarny) [inline src/annotator.cpp]:
```c++
if (((type == CellmlElementType::CONNECTION) || all)
    && Variable::equivalenceConnectionId(v1, v2).empty()) {
```

File: src/annotator.cpp [hunk 8]
```diff
@@ -842,245 +853,159 @@ bool Annotator::assignIds(CellmlElementType type)
...
+                auto id = makeUniqueId();
+                Variable::setEquivalenceConnectionId(v1, v2, id);
+                auto entry = AnyCellmlElement::AnyCellmlElementImpl::create();
+                entry->mPimpl->setConnection(v1, v2);
+                mIdList.insert(std::make_pair(id, convertToWeak(entry)));
+            }
+            if ((type == CellmlElementType::MAP_VARIABLES || all) && Variable::equivalenceMappingId(v1, v2).empty()) {
```
Comment (agarny) [inline src/annotator.cpp]:
```c++
if (((type == CellmlElementType::MAP_VARIABLES) || all)
    && Variable::equivalenceMappingId(v1, v2).empty()) {
```

File: src/annotator.cpp [hunk 9]
```diff
@@ -842,245 +853,159 @@ bool Annotator::assignIds(CellmlElementType type)
...
+                entry->mPimpl->setMapVariables(v1, v2);
+                mIdList.insert(std::make_pair(id, convertToWeak(entry)));
+            }
+        }
     }
-    for (size_t c = 0; c < parent->componentCount(); ++c) {
-        if (parent->component(c)->id().empty()) {
+    for (size_t rIndex = 0; rIndex < component->resetCount(); ++rIndex) {
+        auto r = component->reset(rIndex);
+        if ((type == CellmlElementType::RESET || all) && r->id().empty()) {
```
Comment (agarny) [inline src/annotator.cpp]:
```c++
if (((type == CellmlElementType::RESET) || all)
    && r->id().empty()) {
```

File: src/annotator.cpp [hunk 10]
```diff
@@ -842,245 +853,159 @@ bool Annotator::assignIds(CellmlElementType type)
...
         }
-        doSetComponentIds(parent->component(c));
-    }
-}
-
-void Annotator::AnnotatorImpl::doSetVariableIds(const ComponentPtr &parent)
-{
-    for (size_t v = 0; v < parent->variableCount(); ++v) {
-        if (parent->variable(v)->id().empty()) {
+        if ((type == CellmlElementType::RESET_VALUE || all) && r->resetValueId().empty()) {
```
Comment (agarny) [inline src/annotator.cpp]:
```c++
if (((type == CellmlElementType::RESET_VALUE) || all)
    && r->resetValueId().empty()) {
```

File: src/annotator.cpp [hunk 11]
```diff
@@ -842,245 +853,159 @@ bool Annotator::assignIds(CellmlElementType type)
...
-    for (size_t c = 0; c < parent->componentCount(); ++c) {
-        doSetVariableIds(parent->component(c));
-    }
-}
-
-void Annotator::AnnotatorImpl::doSetResetIds(const ComponentPtr &parent)
-{
-    for (size_t r = 0; r < parent->resetCount(); ++r) {
-        if (parent->reset(r)->id().empty()) {
+        if ((type == CellmlElementType::TEST_VALUE || all) && r->testValueId().empty()) {
```
Comment (agarny) [inline src/annotator.cpp]:
```c++
if (((type == CellmlElementType::TEST_VALUE) || all)
    && r->testValueId().empty()) {
```

**Review by agarny — CHANGES_REQUESTED** [review]

File: src/annotator.cpp (hunk 3 above)
Comment (hsorby) [reply src/annotator.cpp]: [368d482](https://github.com/cellml/libcellml/pull/1151/commits/368d482ef337900a310be0a04b487eb9d2893812)

File: src/annotator.cpp (hunk 4 above)
Comment (hsorby) [reply src/annotator.cpp]: [368d482](https://github.com/cellml/libcellml/pull/1151/commits/368d482ef337900a310be0a04b487eb9d2893812)

File: src/annotator.cpp (hunk 5 above)
Comment (hsorby) [reply src/annotator.cpp]: [368d482](https://github.com/cellml/libcellml/pull/1151/commits/368d482ef337900a310be0a04b487eb9d2893812)

File: src/annotator.cpp (hunk 6 above)
Comment (hsorby) [reply src/annotator.cpp]: [368d482](https://github.com/cellml/libcellml/pull/1151/commits/368d482ef337900a310be0a04b487eb9d2893812)

File: src/annotator.cpp (hunk 7 above)
Comment (hsorby) [reply src/annotator.cpp]: [368d482](https://github.com/cellml/libcellml/pull/1151/commits/368d482ef337900a310be0a04b487eb9d2893812)

File: src/annotator.cpp (hunk 8 above)
Comment (hsorby) [reply src/annotator.cpp]: [368d482](https://github.com/cellml/libcellml/pull/1151/commits/368d482ef337900a310be0a04b487eb9d2893812)

File: src/annotator.cpp (hunk 9 above)
Comment (hsorby) [reply src/annotator.cpp]: [368d482](https://github.com/cellml/libcellml/pull/1151/commits/368d482ef337900a310be0a04b487eb9d2893812)

File: src/annotator.cpp (hunk 10 above)
Comment (hsorby) [reply src/annotator.cpp]: [368d482](https://github.com/cellml/libcellml/pull/1151/commits/368d482ef337900a310be0a04b487eb9d2893812)

File: src/annotator.cpp (hunk 11 above)
Comment (hsorby) [reply src/annotator.cpp]: [368d482](https://github.com/cellml/libcellml/pull/1151/commits/368d482ef337900a310be0a04b487eb9d2893812)

**Review by nickerso — DISMISSED** [review]

File: src/annotator.cpp (hunk 4 above)
Comment (agarny) [reply src/annotator.cpp]: Missing some parentheses:
```c++
bool inHierarchy = (std::dynamic_pointer_cast<libcellml::Model>(component->parent()) == nullptr) || (component->componentCount() > 0);
return ((type == CellmlElementType::COMPONENT_REF) || all) && component->encapsulationId().empty() && inHierarchy;
```

Comment (hsorby) [reply src/annotator.cpp]: [450a656](https://github.com/cellml/libcellml/pull/1151/commits/450a656f9c5d8eb6306d00d05fa57c2d74681be4)

**Review by agarny — DISMISSED** [review]

**Review by nickerso — APPROVED** [review]

**Review by agarny — APPROVED** [review]

**Pull request merged**